Refactor coach and team-assignment logic

nickdnk opened the pull request:
As per discussion with @PhlexPlexico, I've spent some time refactoring the logic for handling coaching. Over the years this code had grown increasingly spaghetti, and a lot of time has been spent by contributors and testers to figure out all the kinks of the coaching system. I've applied everything we learned to this PR in attempt to make the code more readable and easier to understand.

Some key points:

1. The `coach` command in console can no longer be used. We simply block it and tell the client to use the `!coach` chat command or `sm_coach` in console. We do this because the built-in coaching mechanisms are hard to work around and cause weird bugs when used in backups. You can only use this command in freezetime and during warmup.
2. You should now **never** set `players_per_team` to more than the number of players you want to actually have playing. In "scrim mode", anyone joining when the team is full will automatically be placed on coach or kicked if there are no more coach slots. If you are coach and not the last to join, simply call `!coach` in chat and you will be moved there. In non-scrim mode - that is, when both teams are provided in a regular match config - anyone in the `coaches` array will *never* be allowed to join the game as anything but a coach. If you are in the regular `players` array and call `!coach`. you will be moved to the coach array and locked there unless removed using `get5_removeplayer`.
3. You can rejoin the game during warmup if you were not supposed to be coaching. Just call `.coach` again. If the team is full, you cannot rejoin the game.
4. If you join the game during halftime, you are now automatically placed on the right team when the following round starts. This applies to both players and coaches and prevents coaches from simply not selecting a team and ghosting.
5. If your `players_per_team` is set to less than the number of players already on a side, and you load the match configuration, the entire team that exceeded the `players_per_team` will be kicked and must rejoin, but this is a caveat I think we can live with.
6. Dedicated documentation added for coaching to explain all this.
7. Added some defensive if-statements when using `get5_addplayer`,  `get5_addkickedplayer` and `get5_addcoach` to prevent these from being executed during backups or halftime.
8. We **only** update the backup on round start, not when adding or removing players or coaches. A backup should be point-in-time and should not be partially modified at a later stage. If you add a coach or player and restore to a game state where this was not done, they should be kicked from the game and the command should be run again. An exception is the native `SetMatchId` which **has** to update the backup as it changes the match ID.

I've tested this pretty extensively with -maV on Discord and found no issues.

Part 2:
1. Refactored the entire team placement logic. It was a bit of a mess, and this was only because coaching was also a mess.
2. Added a "discouraged parameters" section to the configuration file documentation, as we want to prevent users from putting warmup and restarts into their config files.
3. Got rid of `g_BO2Match` and replace its logic with the added `g_NumberOfMapsInSeries` from https://github.com/splewis/get5/pull/856
4. Fixed a **bunch** of problems with backups, such as being on the wrong side if restoring to a different map and a round number after halftime.
5. Merged `OnConfigsExecuted` and `OnMapStart` since they are always called in succession and SourceMod recommends only using `OnConfigsExecuted` if you set/read cvars: https://sm.alliedmods.net/new-api/sourcemod/OnConfigsExecuted
6. Removed a bunch of `stock` and `public` modifiers.
7. Moved `WriteBackup()` call to `RoundStart` instead of `RoundPreStart`, as it allows for the Valve backup to be written for the first round. For some odd reason, the first round Valve backup has not been written to disk when this call is in `RoundPreStart`, leading to weird handling of valve backups if you wanted to restore to round 0.
8. Added a game restart post-knife, as this is necessary to clear the scoreboard UI. Otherwise the team winning the knife-round will have their win lingering in the middle of the round-win-indicators until the first round is played out. This also helps cleaning up the "going live" logic and tightly couples the announcement to the round start instead of using a timer.
9. Removed a lot of redundant calls to various functions that were stacked on top of eachother.
10. Prevent loading backups during halftime.
11. Added a `RestartGame()` call before loading a Valve backup, as it ensures the game is not in a state that's not prepared for backup (such as bots only alive or after-round).
12. Added pauses used to the backups. They only reset to the backup state if the game is *not* being restored from a live state, meaning they can now be carried over to an entirely new server and restored in a "best effort" manner.
13. We now delete the temporary valve backup file created when exporting the valve backup from a get5 backup.
14. Added `g_MapSides` content to debug output.
15. Fixed missing deletion of "knife decision" timer when a team makes a decision. Previously it would just run its trigger and do nothing because the game was not in a state that expected a knife decision.
16. Fixed some problems with `{MAXMAPS}` and `{MAPNUMBER}` being wrong when formatting them, such as not taking ties or `g_SeriesCanClinch` into consideration.
17. Tightly couple `ExecCfg()` to a call to `SetMatchTeamCvars()` and `ExecuteMatchConfigCvars`, as `ExecCfg` makes a call to `exec xx.cfg` which is async: the two latter must always be set **after** a config file load.
18. Converted `MaxMapsToPlay()` to a more sensible `MapsToWin()` function. The logic was kind of inverted and unnecessary.
19. Fixed a bunch of redundant calls and weird logic in `EnsureIndefiniteWarmup` and merged it into `StartWarmup`.
20. Team-join menu is now never used for any reason.
21. If you disable coaching with `sv_coaching_enabled 0`, existing coaches will be kicked, as they would otherwise become regular spectators. 
22. Added collision protection to the `get5_names.txt` file using the already established logic for this using `{SERVERID}`.

I've tested as much as I can solo, but I need someone to run some tests on this branch.

PhlexPlexico reviewed (commented):
Just a question and some nitpicks for the docs. Otherwise everything else looks good to me/nothing stands out from what I can see!

nickdnk commented on scripting/get5/matchconfig.sp:
```diff
@@ -818,13 +816,16 @@ public Action Command_LoadTeam(int client, int args) {
+    ReplyToCommand(client, "Cannot add players while waiting for round backup.");
+    return Plugin_Handled;
+  } else if (g_PendingSideSwap || InHalftimePhase()) {
+    ReplyToCommand(client, "Cannot add players during halftime. Please wait until the next round starts.");
```
I don't know, but we already block people from joining a team during halftime (if they connect), and I assume there's a good reason for this, so we should also not add coaches or players in that period.

PhlexPlexico replied:
Works for me then!

nickdnk commented on documentation/docs/getting_started.md:
```diff
@@ -18,17 +18,25 @@ call [`get5_creatematch`](../commands/#get5_creatematch). There is also a simple
+## Scrims {: #scrims }
+
+While Get5 is intended for matches (league matches, LAN-matches, cups, etc.), it can be used for everyday
+scrims/gathers/whatever as well. If that is your use case, you should do a few things differently. We call "_having a
```
Sure, this was the original phrasing from the wiki. I'll change it.

nickdnk commented:
Also point #8 in the original comment is important. Did you see that? I added it very recently so not sure if you noticed.

PhlexPlexico commented:
Ah. yeah that would make sense to me at least. I would prefer to have it stored when the coach is added, but it makes sense to have backups as a point in time, so I don't see anything wrong with this change either.

nickdnk commented:
> Ah. yeah that would make sense to me at least. I would prefer to have it stored when the coach is added, but it makes sense to have backups as a point in time, so I don't see anything wrong with this change either.

I realize there's a convenience element to this, but it also introduces a potential for bugs in the backup system that could be super difficult to reproduce and also even fix.

nickdnk commented:
Some additional things I've changed/discovered:

I've disallowed the use of `get5_loadbackup` during halftime. It should probably have been on its own PR, but it's close enough.

You can now only use `.coach` in warmup, similarly to the built-in one. It makes the behavior or scrims vs non-scrim matches more similar, and I don't see any reason why we would ever need to call `.coach` when a game has already started. If you join a full team, you will be put on coach anyway.

You can now **toggle** between coach/player using `.coach` in warmup. Before, you would have to use the team-menu to leave the coach slot.

I've added a "Player is coaching TeamX" message when you coach, as the built-in "player is coaching" message is not used because of the way we place players on coach; it just says "is spectating", which is misleading, so this makes it explicit.

I added a change-hook to `sv_coaching_enabled`, which ensures that all coaches are removed if coaching is disabled, as they will otherwise all be regular spectators, which is *not* good. Obviously, this is a weird-edge case that we should never see under normal circumstances.

The regular team-join menu is now always completely disabled, so you'll never have to use it for any reason.
